specifyweb/frontend/js_src/lib/components/FieldFormatters/FieldFormatter.tsx

```diff
@@ -0,0 +1,183 @@
+import { schemaText } from '../../localization/schema';
+import { getValidationAttributes } from '../../utils/parser/definitions';
+import type { GetSet, RA } from '../../utils/types';
+import { ErrorMessage } from '../Atoms';
+import { Input, Label } from '../Atoms/Form';
+import { ReadOnlyContext } from '../Core/Contexts';
+import type { AnySchema } from '../DataModel/helperTypes';
+import type { SpecifyResource } from '../DataModel/legacyTypes';
+import type { LiteralField } from '../DataModel/specifyField';
+import { softError } from '../Errors/assert';
+import { ResourceMapping } from '../Formatters/Components';
+import { ResourcePreview } from '../Formatters/Preview';
+import { hasTablePermission } from '../Permissions/helpers';
+import type { MappingLineData } from '../WbPlanView/navigator';
+import type { UiFormatter } from '.';
+import { resolveFieldFormatter } from '.';
+import { FieldFormatterParts } from './Parts';
+import type { FieldFormatter } from './spec';
+
+export function FieldFormatterElement({
+  item,
+}: {
+  readonly item: GetSet<FieldFormatter>;
+}): JSX.Element {
+  const [fieldFormatter, setFieldFormatter] = item;
+  const isReadOnly = React.useContext(ReadOnlyContext);
+  return (
+    <>
+      <Label.Inline>
+        <Input.Checkbox
+          checked={fieldFormatter.autoNumber}
+          isReadOnly={isReadOnly}
+          onValueChange={(autoNumber): void =>
+            setFieldFormatter({ ...fieldFormatter, autoNumber })
+          }
+        />
+        {formsText.autoNumbering()}
+      </Label.Inline>
+      <FieldPicker fieldFormatter={item} />
+      {fieldFormatter.external === undefined &&
+      typeof fieldFormatter.table === 'object' ? (
+        <FieldFormatterParts fieldFormatter={item} />
+      ) : (
+        <ErrorMessage>{resourcesText.editorNotAvailable()}</ErrorMessage>
+      )}
+      <FieldFormatterPreview fieldFormatter={fieldFormatter} />
+    </>
+  );
+}
+
+function FieldPicker({
+  fieldFormatter: [fieldFormatter, setFieldFormatter],
+}: {
+  readonly fieldFormatter: GetSet<FieldFormatter>;
+}): JSX.Element | null {
+  const openIndex = React.useState<number | undefined>(undefined);
+  const mapping = React.useMemo(
+    () => (fieldFormatter.field === undefined ? [] : [fieldFormatter.field]),
+    [fieldFormatter.field]
+  );
+  return fieldFormatter.table === undefined ? null : (
+    <Label.Block>
+      {schemaText.field()}
+      <ResourceMapping
+        fieldFilter={excludeNonLiteral}
+        isRequired
+        mapping={[
+          mapping,
+          (mapping): void => {
+            if (mapping !== undefined && mapping?.length > 1)
+              softError('Expected mapping length to be no more than 1');
+            const field = mapping?.[0];
+            if (field?.isRelationship === true) {
+              softError(
+                'Did not expect relationship field in field formatter mapping'
+              );
```

Comment:
these cases should not happen - the UI doesn't allow them. but handling them here just in case (and to satisfy TypeScript)